File: lista1/c/mylib.c
// Marek Traczyński (261748)
// Lista 1 Zadanie 1


#include "mylib.h"


/////////////////////////
// Factorial functions //
/////////////////////////

int my_factorial(int n)
{
    int result = 1;

    if (n < 0 || n > 20)
    {
        return 0;
    }

    for (int i = 2; i <= n; i++)
    {
        result *= i;
    }
    
    return result;
}

int my_factorial_rec(int n)
{
    if (n < 0 || n > 20)
    {
        return 0;
    }

    if (n > 0)
    {
        return n * my_factorial_rec(n - 1);
    }

    return 1;
}


///////////////////
// GCD functions //
///////////////////

int my_gcd(int x, int y)
{
    int temp = 0;

    if (x < 0 || y < 0)
    {
        return 0;

    }

    while (y != 0)
    {
        temp = y;
        y = x % y;
        x = temp;
    }

    return x;
}

int my_gcd_rec(int x, int y)
{
    if (x < 0 || y < 0)
    {
        return 0;
    }    
    
    if (y == 0)
    {
        return x;
    }

    return my_gcd_rec(y, x % y);
}


////////////////////////////////////
// Diophantine equation functions //
////////////////////////////////////

DioGcdResult dio_gcd(int a, int b)
{
    DioGcdResult result = {0, 0, 0};
    int old_rem = a, rem = b;
    int old_x = 1, x = 0;
    int old_y = 0, y = 1;
    int quotient = 0, temp = 0;
    
    while (rem != 0)
    {
        quotient = old_rem / rem;

        temp = old_rem;
        old_rem = rem;
        rem = temp - quotient * rem;
        
        temp = old_x;
        old_x = x;
        x = temp - quotient * x;

        temp = old_y;
        old_y = y;
        y = temp - quotient * y;
    }

    result.g = old_rem;
    result.x = old_x;
    result.y = old_y;

    return result;
}

DioGcdResult dio_gcd_rec(int a, int b)
{
    DioGcdResult result, temp = {0, 0, 0};

    if (b == 0)
    {
        result.g = a;
        result.x = 1;
        result.y = 0;

        return result;
    }

    temp = dio_gcd_rec(b, a % b);
    result.g = temp.g;
    result.x = temp.y;
    result.y = temp.x - (a / b) * temp.y;

    return result;
}

DioResult dio_equation(int a, int b, int c)
{
    DioGcdResult gcd_result = {0, 0, 0};
    DioResult result = {0, 0};

    if (a == 0 && b == 0)
    {
        if (c == 0)
        {
            result.x = 1;
            result.y = 1;
            
            return result;
        }
        else
        {
            result.x = 0;
            result.y = 0;

            return result;
        }
    }

    gcd_result = dio_gcd(a, b);
    
    if (c % gcd_result.g != 0)
    {
        result.x = 0;
        result.y = 0;

        return result;
    }

    result.x = gcd_result.x * (c / gcd_result.g);
    result.y = gcd_result.y * (c / gcd_result.g);

    return result;
}

DioResult dio_equation_rec(int a, int b, int c)
{
    DioGcdResult gcd_result = {0, 0, 0};
    DioResult result = {0, 0};

    if (a == 0 && b == 0)
    {
        if (c == 0)
        {
            result.x = 1;
            result.y = 1;
            
            return result;
        }
        else
        {
            result.x = 0;
            result.y = 0;

            return result;
        }
    }

    gcd_result = dio_gcd_rec(a, b);
    
    if (c % gcd_result.g != 0)
    {
        result.x = 0;
        result.y = 0;

        return result;
    }

    result.x = gcd_result.x * (c / gcd_result.g);
    result.y = gcd_result.y * (c / gcd_result.g);

    return result;
}

File: lista1/julia/mylib.jl
# Marek Traczyński (261748)
# Języki i Paradygmaty Programowania
# Lista 1 Zadanie 3


#########################
## Factorial functions ##
#########################

function myfactorial(n::Integer)::Int
    result::Int = one(Integer)

    if n < 0 || n > 20
        return zero(Integer)
    end

    for i::Int in 2:n
        result *= i
    end
    
    return result
end

function myfactorial_rec(n::Integer)::Int
    if n < 0 || n > 20
        return zero(Integer)
    end

    if n > 0
        return n * myfactorial_rec(n - 1)
    end
    
    return one(Integer)
end


###################
## GCD functions ##
###################

function mygcd(x::Integer, y::Integer)::Int
    if x < 0 || y < 0
        return zero(Integer)
    end

    while y != 0
        (x, y) = (y, x % y)
    end

    return x
end

function mygcd_rec(x::Integer, y::Integer)::Int
    if x < 0 || y < 0
        return zero(Integer)
    end
    
    if y == 0
        return x
    end

    return mygcd_rec(y, x % y)
end


####################################
## Diophantine equation functions ##
####################################

function diogcd(a::Integer, b::Integer)::Tuple{Int, Int, Int}
    (oldrem::Int, rem::Int) = (a, b) 
    (oldx::Int, x::Int) = (one(Integer), zero(Integer)) 
    (oldy::Int, y::Int) = (zero(Integer), one(Integer))
    quotient::Int = zero(Integer)
    
    while rem != 0
        quotient = div(oldrem, rem, RoundDown)
        (oldrem, rem) = (rem, oldrem - quotient * rem)
        (oldx, x) = (x, oldx - quotient * x)
        (oldy, y) = (y, oldy - quotient * y)
    end

    return (oldrem, oldx, oldy)
end

function diogcd_rec(a::Integer, b::Integer)::Tuple{Int, Int, Int}
    if b == 0
        return (a, one(Integer), zero(Integer))
    end

    gcdresult::Tuple{Integer, Integer, Integer} = diogcd_rec(b, a % b)
    g::Int = gcdresult[1]
    x::Int = gcdresult[3]
    y::Int = gcdresult[2] - div(a, b, RoundDown) * gcdresult[3]

    return (g, x, y)
end

function dioequation(a::Integer, b::Integer, c::Integer)::Tuple{Int, Int}
    if a == 0 && b == 0
        if c == 0
            return (one(Integer), one(Integer))
        else
            return (zero(Integer), zero(Integer))
        end
    end

    gcdresult::Tuple{Integer, Integer, Integer} = diogcd(a, b)
    gcdg::Int = gcdresult[1]
    gcdx::Int = gcdresult[2]
    gcdy::Int = gcdresult[3]

    if c % gcdg != 0
        return (zero(Integer), zero(Integer))
    end

    x::Int = gcdx * div(c, gcdg, RoundDown)
    y::Int = gcdy * div(c, gcdg, RoundDown)

    return (x, y)
end

function dioequation_rec(a::Integer, b::Integer, c::Integer)::Tuple{Int, Int}
    if a == 0 && b == 0
        if c == 0
            return (one(Integer), one(Integer))
        else
            return (zero(Integer), zero(Integer))
        end
    end

    gcdresult::Tuple{Integer, Integer, Integer} = diogcd_rec(a, b)
    gcdg::Int = gcdresult[1]
    gcdx::Int = gcdresult[2]
    gcdy::Int = gcdresult[3]

    if c % gcdg != 0
        return (zero(Integer), zero(Integer))
    end

    x::Int = gcdx * div(c, gcdg, RoundDown)
    y::Int = gcdy * div(c, gcdg, RoundDown)

    return (x, y)
end
